rct_image_tools/src/rct_image_tools/charuco_finder.cpp

```diff
@@ -0,0 +1,106 @@
+*/
+
+
+#include <rct_image_tools/charuco.hpp>
+#include <rct_image_tools/charuco_finder.h>
+
+rct_image_tools::CharucoGridBoardObservationFinder::CharucoGridBoardObservationFinder(const cv::Ptr<cv::aruco::CharucoBoard>& board)
+	: board_(board)
+{
+
+}
+
+boost::optional<std::map<int, std::vector<Eigen::Vector2d>>>
+rct_image_tools::CharucoGridBoardObservationFinder::findObservations(const cv::Mat &image) const
+{
+	std::map<int, std::vector<Eigen::Vector2d>> map_ids_to_obs_corners;
+	cv::Ptr<cv::aruco::DetectorParameters> parameters(new cv::aruco::DetectorParameters);
+	std::vector<int> marker_ids;
+    std::vector<std::vector<cv::Point2f> > marker_corners, rejected_candidates;
+    cv::aruco::detectMarkers(image, board_->dictionary, marker_corners, marker_ids, parameters);
+	cv::aruco::refineDetectedMarkers(image, board_, marker_corners, marker_ids, rejected_candidates);
+	for(int i = 0; i < marker_ids.size(); i++)
+	{
+		std::vector<cv::Point2f> corner_pts = marker_corners[i];
+
+		std::vector<Eigen::Vector2d> obs_pts(4);
+		for (int j = 0; j < corner_pts.size(); j++)
+		{
+	  		obs_pts[j] << corner_pts[j].x, corner_pts[j].y;
+		}
+		map_ids_to_obs_corners.insert(std::make_pair(marker_ids[i], obs_pts));
+	}
+	return map_ids_to_obs_corners;	
+}
+
+cv::Mat
+rct_image_tools::CharucoGridBoardObservationFinder::drawObservations(const cv::Mat& image,
+                                                                   const std::map<int, std::vector<Eigen::Vector2d>>& observations) const
+{
+	std::vector<int> marker_ids;
+	std::vector<std::vector<cv::Point2f>> marker_corners;
+	for(std::map<int, std::vector<Eigen::Vector2d>>::const_iterator it = observations.begin(); it != observations.end(); ++it)
+	{
+		marker_ids.push_back(it->first);
+		std::vector<cv::Point2f> cv_obs(it->second.size());
+		std::transform(it->second.begin(), it->second.end(), cv_obs.begin(), [](const Eigen::Vector2d& o) {return cv::Point2d(o.x(), o.y()); });
+		marker_corners.push_back(cv_obs);
+	}
+	
+	cv::aruco::drawDetectedMarkers(image, marker_corners, marker_ids);
+	std::vector<cv::Point2f> charuco_corners;
+	std::vector<int> charuco_ids;
+	cv::aruco::interpolateCornersCharuco(marker_corners, marker_ids, image, board_, charuco_corners, charuco_ids);
+	cv::aruco::drawDetectedCornersCharuco(imageCopy, charuco_corners, charuco_ids, cv::Scalar(255, 0, 0));
+  	return image;
+}
+
+std::map<int, std::vector<Eigen::Vector3d>>
+rct_image_tools::mapCharucoIdsToObjPts(const cv::Ptr<cv::aruco::CharucoBoard> &board)
+{
+  std::map<int, std::vector<Eigen::Vector3d>> map_ids_to_corners;
+  for (int i = 0; i < board->ids.size(); i++)
+  {
+    std::vector<Eigen::Vector3d> obj_pts(board->objPoints[i].size());
+    std::transform(board->objPoints[i].begin(), board->objPoints[i].end(), obj_pts.begin(), [](const cv::Point3f& o) {return Eigen::Vector3d(o.x, o.y, o.z); });
+    map_ids_to_corners.insert(std::make_pair(board->ids[i], obj_pts));
+  }
+  return map_ids_to_corners;
+}
```

Comment:
I would refactor this function to provide a `Correspondence2D3D` type. This is what we require for our calibration problems. I think this method would become:

```c++
std::vector<Correspondence2D3D> createCorrespondences(const cv::Ptr<cv::aruco::CharucoBoard>& board)
```

Reply:
In `aruco_finder.cpp` the purpose of this function is to generate a map between marker IDs and expected target-relative points, which enables lookup of the target points for each visible marker ID. Changing the function to return a vector of correspondences makes it do something fundamentally different than what was intended.

@jaberkebile I think that the way ChArUco grids are detected and represented might be sufficiently different from regular ArUco grids that you might have to fundamentally revisit how you detect observed image points and correlate them to possible target object points. If you can write up a unit test that loads in pictures of ChArUco boards (with some partial views) and uses the detected correlations to perform an extrinsic hand-eye optimization, that should expose these kinds of differences and guide the development of a solution.